// @serverless-devs/s-core/lib/utils/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadTemplateByUrl = exports.proxyDownload = exports.getPackageDownloadUrl = exports.downComponent = exports.writeJsonFile = exports.readJsonFile = exports.generateId = exports.SERVERLESS_GET_APP_INFO_URL = exports.SERVERLESS_API = void 0;
var path = require('path');
var os = require('os');
var axios_1 = require("axios");
var fs = require("fs-extra");
var download = require("download");
exports.SERVERLESS_API = 'https://tool.serverlessfans.com/api';
exports.SERVERLESS_GET_APP_INFO_URL = exports.SERVERLESS_API + '/package/get/object/url';
function generateId() {
    return Date.now();
}
exports.generateId = generateId;
function readJsonFile(filePath) {
    var result = {};
    var data = fs.readFileSync(filePath, 'utf8');
    try {
        result = JSON.parse(data);
    }
    catch (e) {
    }
    return result;
}
exports.readJsonFile = readJsonFile;
function writeJsonFile(filePath, data) {
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf-8');
}
exports.writeJsonFile = writeJsonFile;
function downComponent(componentName, provider, outdir) {
    return __awaiter(this, void 0, void 0, function () {
        var componentDownUrl;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, fs.ensureDir(outdir)];
                case 1:
                    _a.sent();
                    return [4 /*yield*/, getPackageDownloadUrl(componentName, provider)];
                case 2:
                    componentDownUrl = _a.sent();
                    if (!componentDownUrl) return [3 /*break*/, 4];
                    return [4 /*yield*/, downloadTemplateByUrl({ zipFile: componentDownUrl, hasSubPath: false }, outdir)];
                case 3:
                    _a.sent();
                    _a.label = 4;
                case 4: return [2 /*return*/];
            }
        });
    });
}
exports.downComponent = downComponent;
function getPackageDownloadUrl(componentName, provider) {
    return __awaiter(this, void 0, void 0, function () {
        var options, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    options = {
                        url: exports.SERVERLESS_GET_APP_INFO_URL,
                        type: 'get',
                        timeout: 5000,
                        headers: {},
                        params: {
                            name: componentName,
                            provider: provider,
                            type: 'component'
                        }
                    };
                    return [4 /*yield*/, axios_1.default.request(options)];
                case 1:
                    result = _a.sent();
                    if (result.status !== 200) {
                        return [2 /*return*/, null];
                    }
                    if (!result.data.Response) {
                        return [2 /*return*/, null];
                    }
                    return [2 /*return*/, new Promise(function (resolve, reject) {
                            resolve(result.data.Response.Url);
                        })];
            }
        });
    });
}
exports.getPackageDownloadUrl = getPackageDownloadUrl;
function proxyDownload(url, dest, options) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, download(url, dest, { extract: true, strip: 1, timeout: 30000 })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
exports.proxyDownload = proxyDownload;
function downloadTemplateByUrl(template, outputDir) {
    return __awaiter(this, void 0, void 0, function () {
        var uuid, srcDirName, err_1, srcSubDirName, destSubDirName;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    uuid = Date.now();
                    srcDirName = path.join(os.tmpdir(), "" + uuid);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, proxyDownload(template.zipFile, srcDirName, { extract: true, strip: 1 })];
                case 2:
                    _a.sent();
                    return [3 /*break*/, 4];
                case 3:
                    err_1 = _a.sent();
                    return [3 /*break*/, 4];
                case 4:
                    if (!template.hasSubPath) {
                        fs.copySync(srcDirName, outputDir, { dereference: true });
                    }
                    else {
                        srcSubDirName = path.join(srcDirName, template.subPath || '');
                        destSubDirName = path.join(outputDir, template.subPath || '');
                        fs.copySync(srcSubDirName, destSubDirName, { dereference: true });
                    }
                    return [2 /*return*/];
            }
        });
    });
}
exports.downloadTemplateByUrl = downloadTemplateByUrl;
